File: Backend/chat-service/Domain/Common/Models/ApiResponse.cs
namespace ChatService.Application.Common.Models;

public class ApiResponse<T>
{
    public bool Success { get; set; }
    public string Message { get; set; }
    public T Data { get; set; }
    public ErrorDetail Error { get; set; }

    // Helper tạo response thành công nhanh
    public static ApiResponse<T> Ok(T data, string message = "Thành công")
    {
        return new ApiResponse<T>
        {
            Success = true,
            Message = message,
            Data = data,
            Error = null
        };
    }

    // Helper tạo response lỗi
    public static ApiResponse<T> Fail(string message, ErrorDetail error)
    {
        return new ApiResponse<T>
        {
            Success = false,
            Message = message,
            Data = default,
            Error = error
        };
    }
}

public class ErrorDetail
{
    public string Code { get; set; }
    public string Path { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

File: Backend/workspace-service/src/main/java/com/project/workspace_service/shared/response/ApiResponse.java
package com.project.workspace_service.shared.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL) // Bỏ qua field null (để data null thì không hiện, error null không hiện)
public class ApiResponse<T> {

    private boolean success;
    private String message;
    private T data;
    private ErrorDetail error;

    // Class con định nghĩa chi tiết lỗi
    @Data
    @Builder
    public static class ErrorDetail {
        private String code;
        private String path;
        private LocalDateTime timestamp;
    }

    // --- Helper Method: Trả về thành công ---
    public static <T> ApiResponse<T> success(T data, String message) {
        return ApiResponse.<T>builder()
                .success(true)
                .message(message)
                .data(data)
                .error(null)
                .build();
    }

    // --- Helper Method: Trả về lỗi ---
    public static <T> ApiResponse<T> error(String message, String code, String path) {
        return ApiResponse.<T>builder()
                .success(false)
                .message(message)
                .data(null)
                .error(ErrorDetail.builder()
                        .code(code)
                        .path(path)
                        .timestamp(LocalDateTime.now())
                        .build())
                .build();
    }
}